odl/util/vectorization.py

```diff
@@ -124,169 +125,240 @@ def out_shape_from_array(arr):
-                try:
-                    ndim = len(x)
-                except TypeError:
-                    ndim = 1
-
-            if is_valid_input_meshgrid(x, ndim):
-                return _vect_wrapper_meshgrid(x, out, ndim, **kwargs)
-            elif is_valid_input_array(x, ndim):
-                return _vect_wrapper_array(x, out, **kwargs)
-            else:
-                try:
-                    return func(x)
-                except Exception as err:
-                    raise_from(
-                        TypeError('invalid vectorized input type.'), err)
-
-        return vect_wrapper
-    return vect_decorator
+    def decorator(func):
+        if not use_kw:
+            def wrapper(x):
+                if len(x) != nin:
+                    raise TypeError('Need a sequence of length {} as input, '
+                                    'got length {}.'.format(nin, len(x)))
+                return func(*x)
+        else:
+            def wrapper(x, **kwargs):
+                if len(x) != nin:
+                    raise TypeError('Need a sequence of length {} as input, '
+                                    'got length {}.'.format(nin, len(x)))
+                return func(*x, **kwargs)
+
+        return wrapper
+    return decorator
+
+
+class vectorize(object):
+
+    """Decorator class for function vectorization.
+
+    This vectorizer expects a function with exactly one positional
+    argument (input) and optional keyword arguments. The decorated
+    function has an optional ``out`` parameter for in-place evaluation.
+    """
+
+    def __new__(cls, *args, **kwargs):
+        """Create a new decorator class.
+
+        Examples
+        --------
+        Use the decorator witout arguments:
+
+        >>> @vectorize
+        ... def f(x):
+        ...     return x[0] + x[1] if x[0] < x[1] else x[0] - x[1]
+        >>>
+        >>> f([0, 1])  # np.vectorize'd functions always return an array
+        array(1)
+        >>> f([[0, -2], [1, 4]])  # corresponds to points [0, 1], [-2, 4]
+        array([1, 2])
+
+        The function may have ``kwargs``:
+
+        >>> @vectorize
+        ... def f(x, param=1.0):
+        ...     return x[0] + x[1] if x[0] < param else x[0] - x[1]
+        >>>
+        >>> f([[0, -2], [1, 4]])
+        array([1, 2])
+        >>> f([[0, -2], [1, 4]], param=-1.0)
+        array([-1,  2])
+
+        You can pass arguments to the vectorizer, too:
+
+        >>> @vectorize(otypes=['float32'])
+        ... def f(x):
+        ...     return x[0] + x[1] if x[0] < x[1] else x[0] - x[1]
+        >>> f([[0, -2], [1, 4]])
+        array([ 1.,  2.], dtype=float32)
+        """
+        # Decorating without arguments: return wrapper w/o args directly
+        instance = super().__new__(cls)
+
+        if (not kwargs and
+                len(args) == 1 and
+                callable(args[0])):
+            func = args[0]
+
+            return instance._wrapper(func)
```

Comment:
Well, the instance is created before, so it has a `_wrapper` method. It's somewhat dirty since the instance is not yet initialized, so anything relying on a state of the instance would fail. But `_wrapper` doesn't - in fact, it doesn't even use its `self` argument, so the proper solution would be to make it static.